KirchhoffLove.C

```diff
@@ -0,0 +1,258 @@
+// $Id$
+//==============================================================================
+//!
+//! \file KirchhoffLove.C
+//!
+//! \date Sep 13 2011
+//!
+//! \author Knut Morten Okstad / SINTEF
+//!
+//! \brief Base class for linear Kirchhoff-Love thin plate and shell problems.
+//!
+//==============================================================================
+
+#include "KirchhoffLove.h"
+#include "MaterialBase.h"
+#include "ElmMats.h"
+#include "Utilities.h"
+#include "Vec3Oper.h"
+#include "Tensor.h"
+#include "VTF.h"
+
+
+KirchhoffLove::KirchhoffLove (unsigned short int n) : IntegrandBase(n)
+{
+  npv = nsd < 3 ? 1 : 3; // Number of primary unknowns per node
+
+  gravity = 0.0;
+  thickness = 0.1;
+
+  material = nullptr;
+  fluxFld = nullptr;
+  tracFld = nullptr;
+  presFld = nullptr;
+  locSys = nullptr;
+
+  eK = eM = 0;
+  eS = iS = 0;
+}
+
+
+KirchhoffLove::~KirchhoffLove ()
+{
+  if (locSys) delete locSys;
+}
+
+
+void KirchhoffLove::setMode (SIM::SolutionMode mode)
+{
+  m_mode = mode;
+  eM = eK = 0;
+  eS = iS = 0;
+
+  if (mode == SIM::RECOVERY)
+    primsol.resize(1);
+  else
+    primsol.clear();
+
+  switch (mode)
+    {
+    case SIM::STATIC:
+      eK = 1;
+      eS = 1;
+      break;
+
+    case SIM::VIBRATION:
+      eK = 1;
+      eM = 2;
+      break;
+
+    case SIM::STIFF_ONLY:
+      eK = 1;
+      break;
+
+    case SIM::RHS_ONLY:
+      eS = 1;
+      break;
+
+    default:
+      ;
+    }
+}
+
+
+void KirchhoffLove::initIntegration (size_t nGp, size_t nBp)
+{
+  presVal.clear();
+  if (this->haveLoads('I'))
+    presVal.resize(nGp,std::make_pair(Vec3(),Vec3()));
+
+  tracVal.clear();
+  if (this->haveLoads('B'))
+    tracVal.resize(nBp,std::make_pair(Vec3(),Vec3()));
+}
+
+
+LocalIntegral* KirchhoffLove::getLocalIntegral (size_t nen, size_t,
+                                                bool neumann) const
+{
+  ElmMats* result = new ElmMats();
+  switch (m_mode)
+    {
+    case SIM::STATIC:
+      result->rhsOnly = neumann;
+      result->withLHS = !neumann;
+      result->resize(neumann ? 0 : 1, 1);
+      break;
+
+    case SIM::VIBRATION:
+      result->resize(2,0);
+      break;
+
+    case SIM::STIFF_ONLY:
+      result->resize(1,0);
+      break;
+
+    case SIM::RHS_ONLY:
+      result->resize(neumann ? 0 : 1, 1);
+
+    case SIM::RECOVERY:
+      result->rhsOnly = true;
```

Comment:
should there really be no resizing here ?

Reply:
no, in RECOVERY mode we normally don't need any element matrices. Actually, I'm not sure if this method is invoked at all for m_mode == RECOVERY. So we could consider removing this case altogether for the sake of clarity.